File: 061_RotateList.cpp
//11.99%
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
		if(k <= 0 || !head)
			return head;
		
        //First obtain list's length
		int len(1);
		ListNode* curr(head);
		while(curr -> next){
			len++;
			curr = curr -> next;
		}
		
		k %= len;
		if(!k)
			return head;
		curr -> next = head;
		//Find new head of the rotated list
		int index(len - k);
		ListNode* newHead(head);
		while(index){
			curr = newHead;
			newHead = newHead -> next;
			index--;
		}
		curr -> next = NULL;
		return newHead;	
    }
};

//Almost the same idea with mine
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(!head) return head;
        
        int len=1; // number of nodes
        ListNode *newH, *tail;
        newH=tail=head;
        
        while(tail->next)  // get the number of nodes in the list
        {
            tail = tail->next;
            len++;
        }
        tail->next = head; // circle the link

        if(k %= len) 
        {
            for(auto i=0; i<len-k; i++) tail = tail->next; // the tail node is the (len-k)-th node (1st node is head)
        }
        newH = tail->next; 
        tail->next = NULL;
        return newH;
    }
};

File: 061_RotateList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k <= 0:
            return head
        curr, size = head, 0
        while curr:
            size, curr = size + 1, curr.next
        k %= size
        if k == 0:
            return head
        index, newHead = size - k, head
        while index:
            curr, newHead, index = newHead, newHead.next, index - 1
        curr.next = None
        curr = newHead
        while curr.next: curr = curr.next
        curr.next = head
        return newHead
            